File: src/mitris/sim/core/lib/examples/performance/DevStoneGenerator.java
package mitris.sim.core.lib.examples.performance;

import mitris.sim.core.modeling.Atomic;
import mitris.sim.core.modeling.OutPort;

/**
 * Events generator for the DEVStone benchmark
 *
 * @author José Luis Risco Martín
 */
public class DevStoneGenerator extends Atomic {

    public OutPort<Long> oOut = new OutPort<>("out");
    protected double preparationTime;
    protected double period;
    protected long counter = 1;
    protected long maxEvents = Long.MAX_VALUE;

    public DevStoneGenerator(String name, DevStoneProperties properties, long maxEvents) {
        super(name);
        super.addOutPort(oOut);
        this.preparationTime = properties.getPropertyAsDouble(DevStoneProperties.PREPARATION_TIME);
        this.period = properties.getPropertyAsDouble(DevStoneProperties.GENERATOR_PERIOD);
        this.maxEvents = (maxEvents <= 0) ? 1 : maxEvents;
    }

    @Override
    public void initialize() {
        counter = 1;
        this.holdIn("active", preparationTime);
    }

    @Override
    public void deltint() {
        counter++;
        if (counter > maxEvents) {
            super.passivate();
        } else {
            this.holdIn("active", period);
        }
    }

    @Override
    public void deltext(double e) {
        super.passivate();
    }

    @Override
    public void lambda() {
        oOut.addValue(counter);
    }
}

File: c++/src/xdevs/examples/DevStone/DevStoneGenerator.cpp
#include "DevStoneGenerator.h"


DevStoneGenerator::DevStoneGenerator(const std::string& name, double preparationTime, double period, long maxEvents)
	: Atomic(name),
	  preparationTime(preparationTime),
	  period(period),
	  maxEvents(maxEvents),
	  counter(0),
	  oOut("out") {
	Component::addOutPort(&oOut);
}

DevStoneGenerator::~DevStoneGenerator() { }

void DevStoneGenerator::initialize() {
	this->counter = 1;
	Atomic::holdIn("active", preparationTime);
}

void DevStoneGenerator::exit() { }

void DevStoneGenerator::deltint() {
	counter++;
	if (counter > maxEvents) {
		Atomic::passivate();
	} else {
		Atomic::holdIn("active", period);
	}
}

void DevStoneGenerator::deltext(double e) {
	Atomic::passivate();
}

void DevStoneGenerator::lambda() {
	Event event = Event::makeEvent(new long(counter));
	oOut.addValue(event);
}
